applications/StructuralMechanicsApplication/custom_constitutive/viscous_generalized_kelvin_3d.h

```diff
@@ -0,0 +1,363 @@
+        this->SetNonConvPreviousInelasticStrainVector(InelasticStrainVector);
+
+    } // End CalculateMaterialResponseCauchy
+
+    void FinalizeSolutionStep(
+        const Properties& rMaterialProperties,
+        const GeometryType& rElementGeometry,
+        const Vector& rShapeFunctionsValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override
+    {
+        // Update the required vectors
+        this->SetPreviousInelasticStrainVector(this->GetNonConvPreviousInelasticStrainVector());
+        this->SetPreviousStressVector(this->GetNonConvPreviousStressVector());
+    }
+
+    void CalculateElasticMatrix(Matrix &rElasticityTensor,
+        const Properties &rMaterialProperties
+    )
+    {
+        const double E = rMaterialProperties[YOUNG_MODULUS];
+        const double poisson_ratio = rMaterialProperties[POISSON_RATIO];
+        const double lambda =
+            E * poisson_ratio / ((1. + poisson_ratio) * (1.0 - 2.0 * poisson_ratio));
+        const double mu = E / (2.0 + 2.0 * poisson_ratio);
+
+        if (rElasticityTensor.size1() != 6 || rElasticityTensor.size2() != 6)
+            rElasticityTensor.resize(6, 6, false);
+        rElasticityTensor.clear();
+
+        rElasticityTensor(0, 0) = lambda + 2.0 * mu;
+        rElasticityTensor(0, 1) = lambda;
+        rElasticityTensor(0, 2) = lambda;
+        rElasticityTensor(1, 0) = lambda;
+        rElasticityTensor(1, 1) = lambda + 2.0 * mu;
+        rElasticityTensor(1, 2) = lambda;
+        rElasticityTensor(2, 0) = lambda;
+        rElasticityTensor(2, 1) = lambda;
+        rElasticityTensor(2, 2) = lambda + 2.0 * mu;
+        rElasticityTensor(3, 3) = mu;
+        rElasticityTensor(4, 4) = mu;
+        rElasticityTensor(5, 5) = mu;
+    }
+
+    void FinalizeMaterialResponsePK1(ConstitutiveLaw::Parameters& rValues)
+    {
+    }
+    void FinalizeMaterialResponsePK2(ConstitutiveLaw::Parameters& rValues)
+    {
+    }
+    void FinalizeMaterialResponseKirchhoff(ConstitutiveLaw::Parameters& rValues)
+    {
+    }
+    void FinalizeMaterialResponseCauchy(ConstitutiveLaw::Parameters& rValues)
+    {
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+    ///@}
+
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+private:
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // Converged values
+    Vector mPrevStressVector = ZeroVector(6);
```

Comment:
do you actually need both of them? i am asking since i would guess with one you could have enough (but admittedly i may well be wrong about this)

Reply:
in this case, the viscous CL need te previous StressVector so it's strictly necessary to store them...